src/pudl/helpers.py

```diff
@@ -332,90 +332,271 @@ def is_doi(doi):
     return bool(re.match(doi_regex, doi))
 
 
-def clean_merge_asof(
+def convert_col_to_datetime(df, date_col_name):
+    """Convert a column in a dataframe to a datetime.
+
+    If the column isn't a datetime, it needs to be converted to a string type
+    first so that integer years are formatted correctly.
+
+    Args:
+        df (pandas.DataFrame): Dataframe with column to convert.
+        date_col_name (string): name of the column to convert.
+
+    Returns:
+        Dataframe with the converted datetime column.
+    """
+    if pd.api.types.is_datetime64_ns_dtype(df[date_col_name]) is False:
+        logger.warning(
+            f"{date_col_name} is {df[date_col_name].dtype} column. Converting to datetime."
+        )
+        df[date_col_name] = pd.to_datetime(df[date_col_name].astype("string"))
+    return df
+
+
+def full_timeseries_date_merge(
     left: pd.DataFrame,
     right: pd.DataFrame,
-    left_on: str = "report_date",
-    right_on: str = "report_date",
-    by: List[str] = [],
+    on: List[str],
+    left_date_col: str = "report_date",
+    right_date_col: str = "report_date",
+    new_date_col: str = "report_date",
+    date_on: List[str] = ["year"],
+    how: Literal["inner", "outer", "left", "right", "cross"] = "inner",
+    report_at_start: bool = True,
+    freq: str = "MS",
+    **kwargs,
+):
+    """Merge dataframes with different date frequencies and expand to a full timeseries.
+
+    Arguments: see arguments for ``date_merge`` and ``expand_timeseries``
+    """
+    out = date_merge(
+        left=left,
+        right=right,
+        left_date_col=left_date_col,
+        right_date_col=right_date_col,
+        new_date_col=new_date_col,
+        on=on,
+        date_on=date_on,
+        how=how,
+        report_at_start=report_at_start,
+        **kwargs,
+    )
+    out = expand_timeseries(
+        df=out,
+        date_col=new_date_col,
+        freq=freq,
+        key_cols=on,
+    )
+    return out
+
+
+def _add_suffix_to_date_on(date_on):
+    """Check date_on list is valid and add _temp_for_merge suffix."""
+    if date_on is None:
+        date_on = ["year"]
+    date_on_suffix = []
+    for col in date_on:
+        if col not in ["year", "month", "quarter", "day"]:
+            raise AssertionError(
+                logger.error(f"{col} is not a valid string in date_on column list.")
+            )
```

Comment:
This assert was put in in response to the issue with type hints and `Literal["year", "month", "quarter", "day"]` (see [this thread](https://github.com/catalyst-cooperative/pudl/pull/1550#discussion_r883874719)).